README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://bioconductor.org/packages/aroma.light/"><img border="0" src="https://bioconductor.org/shields/years-in-bioc/aroma.light.svg" alt="Bioconductor since badge"/></a> <a href="https://bioconductor.org/checkResults/release/bioc-LATEST/aroma.light/"><img border="0" src="https://bioconductor.org/shields/build/release/bioc/aroma.light.svg" alt="Bioconductor build status"/></a> <a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/aroma.light/"><img border="0" src="https://bioconductor.org/shields/build/devel/bioc/aroma.light.svg" alt="Bioconductor build status"/></a> <a href="https://github.com/HenrikBengtsson/aroma.light/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/aroma.light/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>     <a href="https://app.codecov.io/gh/HenrikBengtsson/aroma.light"><img border="0" src="https://codecov.io/gh/HenrikBengtsson/aroma.light/branch/develop/graph/badge.svg" alt="Coverage Status"/></a> 
</div>

# aroma.light: Light-Weight Methods for Normalization and Visualization of Microarray Data using Only Basic R Data Types 


## Installation
R package aroma.light is available on [Bioconductor](https://www.bioconductor.org/packages/devel/bioc/html/aroma.light.html) and can be installed in R as:

```r
install.packages("BiocManager")
BiocManager::install("aroma.light")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/aroma.light", ref="develop")
```
This will install the package from source.  

<!-- pkgdown-drop-below -->


## Contributing

To contribute to this package, please see [CONTRIBUTING.md](CONTRIBUTING.md).



NEWS
Package: aroma.light
====================

Version 3.29.0 [2022-11-01]

 * The version number was bumped for the Bioconductor devel version,
   which is now Bioconductor 3.17 for R devel.


Version 3.28.0 [2022-11-01]

 * The version number was bumped for the Bioconductor release version,
   which is now Bioconductor 3.16 for R (>= 4.2.2).


Version: 3.27.0 [2022-04-26]

 * The version number was bumped for the Bioconductor devel version,
   which is now BioC 3.16 for R-devel.


Version: 3.26.0 [2022-04-26]

 * The version number was bumped for the Bioconductor release version,
   which is now BioC 3.15 for R (>= 4.2.0).


Version: 3.25.1 [2022-04-21]

BUG FIXES:

 * Fixed partial argument name used in iwpca().
 

Version: 3.24.0 [2021-10-27]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.14 for R (>= 4.1.1).


Version: 3.23.1 [2021-08-19]

DOCUMENTATION:

 * Update several citation URLs that were either broken or redirects elsewhere.


Version: 3.22.0 [2021-05-19]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.13 for R (>= 4.0.3).


Version: 3.20.0 [2020-10-27]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.12 for R (>= 4.0.0).


Version: 3.18.0 [2020-04-27]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.11 for R (>= 3.6.1).


Version: 3.17.1 [2019-12-09]

CODE REFACTORING:

 * Now importing throw() from R.oo instead of R.methodsS3.


Version: 3.17.0 [2019-10-30]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.11 for R (>= 3.6.1).


Version: 3.16.0 [2019-10-30]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.10 for R (>= 3.6.1).


Version: 3.15.1 [2019-08-28]

BUG FIXES:

 * wpca() for matrices had a 'length > 1 in coercion to logical' bug.


Version: 3.15.0 [2019-05-02]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.10 for R (>= 3.6.0).


Version: 3.14.0 [2019-05-02]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.9 for R (>= 3.6.0).


Version: 3.13.0 [2018-10-30]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.9 for R (>= 3.6.0).


Version: 3.12.0 [2018-10-30]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.8 for R (>= 3.5.0).


Version: 3.11.2 [2018-09-04]

CODE REFACTORING:

 * fitPrincipalCurve() now requires princurve (>= 2.1.2) and was updated to
   make use of new principcal_curve class instead of deprecated
   principcal.curve class.  This update "should not" affect the results, but
   see https://github.com/dynverse/princurve/issues/8 for information of what
   has changed in the princurve package in this respect.
 

Version: 3.11.1 [2018-08-28]

 * Updated installation instructions in README.md.


Version: 3.11.0 [2018-04-30]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.8 for R (>= 3.6.0).


Version: 3.10.0 [2018-04-30]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.7 for R (>= 3.5.0).
 

Version: 3.9.1 [2017-12-19]

NEW FEATURES:

 * robustSmoothSpline() now supports using Tukey's biweight (in addition to
   already exising L1) estimators.  See argument 'method'.  Thanks to
   Aaron Lun at the Cancer Research UK Cambridge Institute for adding this
   feature.


Version: 3.9.0 [2017-10-30]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.7 for R (>= 3.5.0).


Version: 3.8.0 [2017-10-30]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.6 for R (>= 3.4.0).


Version: 3.7.0 [2017-04-23]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC 3.6 for R (>= 3.4.0).


Version: 3.6.0 [2017-04-23]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC 3.5 for R (>= 3.4.0).


Version: 3.5.1 [2017-04-14]
 
SIGNIFICANT CHANGES:

 * robustSmoothSpline() uses a re-weighted re-iterative algorithm that fits a
   smooth spline using stats::smooth.spline(), calculates the residuals and
   which are used to fit a re-weighted smooth spline and so on until converence.
   Due to updates to stats::smooth.spline() in R (>= 3.4.0) it is no longer
   feasible to maintain a highly optimized version of the algorithm, because it
   was based on internal stats::smooth.spline() code that has no completely
   changed.  Instead the re-iterative algorithm calls stats::smooth.spline() as
   is, which slows it down.  More importantly, it will now give slightly
   different estimates.

SOFTWARE QUALITY:

 * In addition to continous integration (CI) tests and nightly Bioconductor
   tests, the package is now also tested regularly against all reverse package
   depencies available on CRAN and Bioconductor.
 
 
Version: 3.5.0 [2016-10-18]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.5 for R (>= 3.4.0).
 
 
Version: 3.4.0 [2016-10-18]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.4 for R (>= 3.3.1).
 
 
Version: 3.3.2 [2016-09-16]

CODE REFACTORING:

 * Using NA_real_ (not just NA) everywhere applicable.

BUG FIXES:

 * robustSmoothSpline() gave an error since R-devel (>= 3.4.0 r70682).
 
 
Version: 3.3.1 [2016-08-10]

CODE REFACTORING:

 * CLEANUP: Using seq_len() and seq_along() everywhere (Issue #8)
 
 
Version: 3.3.0 [2016-05-03]

 * The version number was bumped for the Bioconductor devel version, which is
   now BioC v3.4 for R (>= 3.3.0).
 
 
Version: 3.2.0 [2016-05-03]

 * The version number was bumped for the Bioconductor release version, which is
   now BioC v3.3 for R (>= 3.3.0).
 
 

R/iwpca.R
#########################################################################/**
# @RdocGeneric iwpca
# @alias iwpca.matrix
#
# @title "Fits an R-dimensional hyperplane using iterative re-weighted PCA"
#
# \description{
#   @get "title".
# }
#
# \usage{
# @usage iwpca,matrix
# }
#
# \arguments{
#  \item{X}{N-times-K @matrix where N is the number of observations and
#    K is the number of dimensions.}
#  \item{w}{An N @vector of weights for each row (observation) in
#    the data matrix. If @NULL, all observations get the same weight.}
#  \item{R}{Number of principal components to fit. By default a line
#    is fitted.}
#  \item{method}{
#    If \code{"symmetric"} (or \code{"bisquare"}), Tukey's biweight
#    is used. If \code{"tricube"}, the tricube weight is used.
#    If \code{"L1"}, the model is fitted in \eqn{L_1}.
#    If a @function, it is used to calculate weights for next iteration
#    based on the current iteration's residuals.}
#  \item{maxIter}{Maximum number of iterations.}
#  \item{acc}{The (Euclidean) distance between two subsequent parameters
#    fit for which the algorithm is considered to have converged.}
#  \item{reps}{Small value to be added to the residuals before the
#    the weights are calculated based on their inverse. This is to avoid
#    infinite weights.}
#  \item{fit0}{A @list containing elements \code{vt} and \code{pc}
#    specifying an initial fit.
#    If @NULL, the initial guess will be equal to the (weighted) PCA fit.}
#  \item{...}{Additional arguments accepted by @see "wpca".}
# }
#
# \value{
#   Returns the fit (a @list) from the last call to @see "wpca"
#   with the additional elements \code{nbrOfIterations} and
#   \code{converged}.
# }
#
# \details{
#   This method uses weighted principal component analysis (WPCA) to fit a
#   R-dimensional hyperplane through the data with initial internal
#   weights all equal.
#   At each iteration the internal weights are recalculated based on
#   the "residuals".
#   If \code{method=="L1"}, the internal weights are 1 / sum(abs(r) + reps).
#   This is the same as \code{method=function(r) 1/sum(abs(r)+reps)}.
#   The "residuals" are orthogonal Euclidean distance of the principal
#   components R,R+1,...,K.
#   In each iteration before doing WPCA, the internal weighted are
#   multiplied by the weights given by argument \code{w}, if specified.
# }
#
# @author
#
# @examples "../incl/iwpca.matrix.Rex"
#
# \seealso{
#   Internally @see "wpca" is used for calculating the weighted PCA.
# }
#
# @keyword "algebra"
#*/#########################################################################
setMethodS3("iwpca", "matrix", function(X, w=NULL, R=1, method=c("symmetric", "bisquare", "tricube", "L1"), maxIter=30, acc=1e-4, reps=0.02, fit0=NULL, ...) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 1. Verify the arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument: 'w'
  if (!is.null(w))
    w <- rep(w, length.out=nrow(X));
  w0 <- w;

  # Argument: 'method'
  if (is.function(method)) {
    dummy <- method(1:5);
    if (!is.numeric(dummy))
      stop("Argument 'method' (weight function) does not return numeric values.");
    if (!is.vector(dummy))
      stop("Argument 'method' (weight function) does not return a vector.");
    if (length(dummy) != 5)
      stop("Argument 'method' (weight function) does not return the correct number of values.");
  } else {
    method <- match.arg(method);
  }

  # Argument: 'fit0'
  if (!is.null(fit0)) {
    if (!all(c("vt", "pc") %in% names(fit0))) {
      throw("Argument 'fit0' is missing element 'vt' or 'pc': ",
                                        paste(names(fit0), collapse=", "));
    }
  }


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 2. Fit the model
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Ulast <- 1/.Machine$double.eps; # A large number
  iter <- 0;
  isConverged <- FALSE;
  w <- rep(1, times=nrow(X));
  while (!isConverged && iter < maxIter) {
    if (iter > 0 || is.null(fit0)) {
      iter <- iter + 1;

      # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      # "Re-weight the weights"
      # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      if (!is.null(w0))
        w <- w0 * w;

      # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      # Fit N-dimensional weighted PCA.
      # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      fit <- wpca(X, w=w, scale=FALSE, ...);
    } else {
      fit <- fit0;
    }

    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    # Get the fitted line L
    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    # Get fitted eigenvectors (u1,u2,...,uN)
    # with ui*uj = 0; i!=j and ui*ui = 1.
    U <- fit$vt;
    colnames(U) <- rownames(U) <- NULL;

    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    # Check for convergence
    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    isConverged <- (sum(abs(U-Ulast))/length(U) < acc);
    Ulast <- U;

    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    # Finally, update the weights
    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    # Residuals in the "tailing" dimensions.
    r <- fit$pc[,-c(1:R), drop=FALSE];

    # Residuals in orthogonal Euclidean distance
    if (anyMissing(r)) {
      # Sometimes some residuals become NAs.
      r <- sqrt(rowSums(r^2, na.rm=TRUE));
    } else {
      r <- sqrt(rowSums(r^2));
    }

    # Down-weight points that are "far" away...
    if (is.character(method)) {
      if (method == "L1") {
        # Add small number to residuals to avoid infinite weights
        r <- abs(r) + reps;
        w <- 1/r;
      } else if (method %in% c("symmetric", "bisquare")) {
        # Add small number to residuals to avoid infinite weights
        r <- abs(r) + reps;
        r <- r/6;
        # Zero weights introduce NA's (for unknown reasons), therefore
        # with use a number very close to zero instead.
        w <- rep(.Machine$double.eps, times=length(r));
        ii <- (r < 1);
        w[ii] <- (1-r[ii]^2)^2;
        # Not needed anymore
        ii <- NULL;
      } else if (method == "tricube") {
        # Add small number to residuals to avoid infinite weights
        r <- abs(r) + reps;
        r <- r/6;
        # Zero weights introduce NA's (for unknown reasons), therefore
        # with use a number very close to zero instead.
        w <- rep(.Machine$double.eps, times=length(r));
        ii <- (r < 1);
        w[ii] <- (1-r[ii]^3)^3;
        # Not needed anymore
        ii <- NULL;
      }
    } else if (is.function(method)) {
      # Pass also the "fitted values" to the weight function.
      attr(r, "x") <- fit$pc[,1:R, drop=FALSE];
      attr(r, "reps") <- reps;
      w <- method(r);

      # Combine w_i = ||w_{i,j}||_2 (Euclidean distance), if needed.
      if (is.matrix(w) && ncol(w) > 1) {
        w <- sqrt(rowSums(w^2)/ncol(w));
      } else {
        w <- as.vector(w);
      }
    }

    # Sometimes some weights become NAs (not sure why). Set them to zero.
    if (anyMissing(w)) {
      w[is.na(w)] <- 0;
    }

    # Not needed anymore
    r <- NULL;
  } # while(...)

  fit$w <- w;
  fit$nbrOfIterations <- iter;
  fit$converged <- isConverged;

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 3. Return the estimated parameters
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  fit;
}) # iwpca()


############################################################################
# HISTORY:
# 2013-09-26
# o Now utilizing anyMissing().
# 2006-04-25
# o Updated the example to first plot data from all viewpoints, then just
#   the lines.  Faster since the lines are only fitted once and nicer.
# 2005-05-03
# o Now test of argument 'fit0' checks if it is NULL.
# 2005-03-28
# o Second try with initial guess; added argument 'fit0'.
# 2005-02-08
# o Added "symmetric" (now default) and "tricube" too.
# 2005-02-07
# o Argument 'method' is now how the weights are calculated from the
#   residuals, not how residuals are combined across dimensions.
#   Method "L2" is therefore removed, because it corresponds to wpca().
# o Added support for weight functions via argument 'method'. The function
#   must take a matrix of residuals as the first argument.
# 2004-05-14
# o Made into a method of class matrix instead of a stand-alone function.
# 2004-04-18
# o Added support for weighted IWPCA; simply by weighing the weights.
# 2004-03-09
# o Now making use of our own weighted PCA method instead of the
#   multidim::acp() method, which is not maintained anymore.
# 2003-12-29
# o Generalized the method to fit a R-dimensional hyperplane instead of
#   just a line, which is the default fit.
# 2003-12-28
# o Added to the R.basic package.
# o Created by extracted code from RGData.fitMultiIWPCA().
############################################################################
